File: 백준/Gold/9663. N－Queen/N－Queen.cc
#include <bits/stdc++.h>

using namespace std;

int n;
bool isused1[30];
bool isused2[30];
bool isused3[30];

int cnt = 0;

void func(int cur) {
	if (cur == n) {
		cnt++;
		return;
	}

	for (int i = 0; i < n; i++) {
		if (!isused1[i] && !isused2[cur + i] && !isused3[cur - i + n - 1]) {
			isused1[i] = 1;
			isused2[cur + i] = 1;
			isused3[cur - i + n - 1] = 1;
			func(cur + 1);
			isused1[i] = 0;
			isused2[cur + i] = 0;
			isused3[cur - i + n - 1] = 0;

		}
	}
}
int main(void) {
	ios::sync_with_stdio(0);
	cin.tie(0);

	cin >> n;
	func(0);

	cout << cnt;
}

File: 백준/Gold/9663. N－Queen/N－Queen.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Main {

	static int N, answer;
	static boolean[] isused1 = new boolean[20];
	static boolean[] isused2 = new boolean[50];
	static boolean[] isused3 = new boolean[50];
	
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	
		N = Integer.parseInt(br.readLine());
		
		backTracking(0);
		
		System.out.println(answer);
	}
	
	static void backTracking(int cur) {
		if(cur == N) {
			answer++;
			return;
		}
		
		
		for(int i = 0; i < N; i++) {
			if(isused1[i] || isused2[cur + i] || isused3[cur - i + N - 1]) continue;
			
			isused1[i] = true;
			isused2[cur + i] = true;
			isused3[cur - i + N - 1] = true;
			backTracking(cur + 1);
			isused1[i] = false;
			isused2[cur + i] = false;
			isused3[cur - i + N - 1] = false;
			
		}
	}
	
	
}

